File: 2023/java/day04a/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashSet;

public class Main {
    static void usage() {
        System.out.println("usage: java Main <file>");
        System.exit(1);
    }

    static int process(String filename) {
        int result = 0;
        try {
            BufferedReader reader = new BufferedReader(new FileReader(filename));
            String line;
            while ( (line = reader.readLine()) != null ) {
                String rest = line.split(":\\s+")[1];
                String winningStr = rest.split("\\s+\\|\\s+")[0];
                HashSet<Integer> winningSet = new HashSet<Integer>();
                for ( String winningNum : winningStr.split("\\s+") ) {
                    winningSet.add(Integer.parseInt(winningNum));
                }
                String handStr = rest.split("\\s+\\|\\s+")[1];
                HashSet<Integer> handSet = new HashSet<Integer>();
                for ( String handNum : handStr.split("\\s+") ) {
                    handSet.add(Integer.parseInt(handNum));
                }
                HashSet<Integer> intersection = new HashSet<Integer>(winningSet);
                intersection.retainAll(handSet);
                result += (intersection.size() > 0) ? 1 << (intersection.size() - 1) : 0;
            }
            reader.close();
        } catch ( IOException e ) {
            e.printStackTrace();
        }
        return result;
    }

    public static final void main(String[] args) {
        if ( args.length < 1 ) {
            usage();
        }
        String filename = args[0];
        int result = process(filename);
        System.out.println("result = " + Integer.toString(result));
    }
}

File: 2023/haskell/day04a/Main.hs
module Main ( main ) where

import Data.Char
import qualified Data.Set as Set
import Data.Maybe
import System.Environment
import System.Exit
import System.IO

splitWhen     :: (Char -> Bool) -> String -> [String]
splitWhen pred s =  case dropWhile pred s of
                      "" -> []
                      s' -> w : splitWhen pred s''
                            where (w, s'') = break pred s'

process :: String -> Int
process contents =
  foldl
    (\result line ->
      let rest = head $ drop 1 $ splitWhen (== ':') line
          winningStr = head $ splitWhen (== '|') rest
          winningSet = Set.fromList $ (map read (words winningStr) :: [Int])
          handStr = head $ drop 1 $ splitWhen (== '|') rest
          handSet = Set.fromList $ (map read (words handStr) :: [Int])
          commonCount = length (Set.intersection winningSet handSet)
      in result + (if (commonCount > 0) then (2 ^ (commonCount - 1)) else 0)
    )
    0
    (lines contents)

usage :: IO a
usage = do
  progname <- getProgName
  putStrLn $ "usage: " ++ progname ++ " <file>"
  exitWith (ExitFailure 1)

main :: IO ()
main = do
  args <- getArgs
  case args of
    [] -> usage
    [filename] -> do
      contents <- readFile filename
      let result = process contents
      putStrLn $ "result = " ++ show result
    _ -> usage
